[rtsan] Error while building `rtsan_interceptors.cpp`

We've been seeing this issue since [2024-07-23](https://github.com/fedora-llvm-team/llvm-snapshots/issues/604).

That makes me believe this started to happen after b177ac4a44ec4024fec6927f4870ee96e9070696 .

Error message:
```
FAILED: compiler-rt/lib/rtsan/CMakeFiles/RTRtsan.x86_64.dir/rtsan_interceptors.cpp.o 
/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/llvm/redhat-linux-build/./bin/clang++ --target=x86_64-redhat-linux-gnu -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/.. -O2 -flto=thin -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS --config=/usr/lib/rpm/redhat/redhat-hardened-clang.cfg -fstack-protector-strong -m64 -march=x86-64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_DEFAULT_SOURCE -Dasm=__asm__ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O2 -g -DNDEBUG -std=c++17 -m64 -fPIC -fno-builtin -fno-exceptions -fomit-frame-pointer -funwind-tables -fno-stack-protector -fno-sanitize=safe-stack -fvisibility=hidden -fno-lto -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -O3 -gline-tables-only -Wno-gnu -Wno-variadic-macros -Wno-c99-extensions -ftrivial-auto-var-init=pattern -DSANITIZER_COMMON_NO_REDEFINE_BUILTINS -MD -MT compiler-rt/lib/rtsan/CMakeFiles/RTRtsan.x86_64.dir/rtsan_interceptors.cpp.o -MF compiler-rt/lib/rtsan/CMakeFiles/RTRtsan.x86_64.dir/rtsan_interceptors.cpp.o.d -o compiler-rt/lib/rtsan/CMakeFiles/RTRtsan.x86_64.dir/rtsan_interceptors.cpp.o -c /builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/rtsan_interceptors.cpp
/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:388:22: error: address of overloaded function 'openat' does not match required type 'unsigned long'
  388 |   INTERCEPT_FUNCTION(openat);
      |                      ^~~~~~
/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/../interception/interception.h:371:71: note: expanded from macro 'INTERCEPT_FUNCTION'
  371 | # define INTERCEPT_FUNCTION(func) INTERCEPT_FUNCTION_LINUX_OR_FREEBSD(func)
      |                                                                       ^~~~
/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/../interception/interception_linux.h:35:33: note: expanded from macro 'INTERCEPT_FUNCTION_LINUX_OR_FREEBSD'
   35 |       (::__interception::uptr)&(func),            \
      |                                 ^~~~
/builddir/build/BUILD/llvm-19.0.0_pre20240723.g40954d7f9bb38b-build/llvm-project-40954d7f9bb38b2407fe48a524befc5216f13ccc/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:72:18: note: candidate function
   72 | INTERCEPTOR(int, openat, int fd, const char *path, int oflag, ...) {
      |                  ^
/usr/include/bits/fcntl2.h:195:1: note: candidate address cannot be taken because parameter 2 has pass_object_size attribute
  195 | openat (int __fd, __fortify_clang_overload_arg (const char *, ,__path),
      | ^
/usr/include/bits/fcntl2.h:185:1: note: candidate address cannot be taken because parameter 2 has pass_object_size attribute
  185 | openat (int __fd, __fortify_clang_overload_arg (const char *, ,__path),
      | ^
/usr/include/bits/fcntl2.h:181:1: note: candidate function
  181 | openat (int __fd, const char *__path, int __oflag, mode_t __mode, ...)
      | ^
1 error generated.
```

The full build log is available at: [builder-live.log.gz](https://github.com/user-attachments/files/16393123/builder-live.log.gz)

--- a/compiler-rt/lib/interception/interception_linux.h
+++ b/compiler-rt/lib/interception/interception_linux.h
@@ -28,12 +28,14 @@ bool InterceptFunction(const char *name, const char *ver, uptr *ptr_to_real,
                        uptr func, uptr trampoline);
 }  // namespace __interception
 
-#define INTERCEPT_FUNCTION_LINUX_OR_FREEBSD(func) \
-  ::__interception::InterceptFunction(            \
-      #func,                                      \
-      (::__interception::uptr *)&REAL(func),      \
-      (::__interception::uptr)&(func),            \
-      (::__interception::uptr)&TRAMPOLINE(func))
+// Cast func to type of REAL(func) before casting to uptr in case it is an
+// overloaded function, which is the case for some glibc functions when
+// _FORTIFY_SOURCE is used. This disambiguates which overload to use.
+#define INTERCEPT_FUNCTION_LINUX_OR_FREEBSD(func)            \
+  ::__interception::InterceptFunction(                       \
+      #func, (::__interception::uptr *)&REAL(func),          \
+      (::__interception::uptr)(decltype(REAL(func)))&(func), \
+      (::__interception::uptr) &TRAMPOLINE(func))
 
 // dlvsym is a GNU extension supported by some other platforms.
 #if SANITIZER_GLIBC || SANITIZER_FREEBSD || SANITIZER_NETBSD
@@ -41,7 +43,7 @@ bool InterceptFunction(const char *name, const char *ver, uptr *ptr_to_real,
   ::__interception::InterceptFunction(                        \
       #func, symver,                                          \
       (::__interception::uptr *)&REAL(func),                  \
-      (::__interception::uptr)&(func),                        \
+      (::__interception::uptr)(decltype(REAL(func)))&(func),  \
       (::__interception::uptr)&TRAMPOLINE(func))
 #else
 #define INTERCEPT_FUNCTION_VER_LINUX_OR_FREEBSD(func, symver) \

